// @atomist/automation-client/cli/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const GitHubApi = require("@octokit/rest");
const inquirer = require("inquirer");
const stringify = require("json-stringify-safe");
const _ = require("lodash");
const os = require("os");
const configuration_1 = require("../configuration");
const github = new GitHubApi();
function createGitHubToken(user, password, mfa) {
    github.authenticate({
        type: "basic",
        username: user,
        password,
    });
    const host = os.hostname();
    const params = {
        scopes: ["read:org", "repo"],
        note: `Atomist API on ${host}`,
        note_url: "http://www.atomist.com/",
    };
    if (mfa) {
        params.headers = { "X-GitHub-OTP": mfa };
    }
    return github.authorization.create(params).then(res => {
        if (res.data && res.data.token) {
            return res.data.token;
        }
        throw new Error(`GitHub API returned successful but there is no token`);
    });
}
function cliAtomistConfig(argv) {
    const argTeamId = argv.team;
    const argGitHubUser = argv["github-user"];
    const argGitHubPassword = argv["github-password"];
    const argGitHubMfaToken = argv["github-mfa-token"];
    const userConfig = configuration_1.getUserConfig() || {};
    if (!userConfig.teamIds) {
        userConfig.teamIds = [];
    }
    if (argTeamId) {
        if (!userConfig.teamIds.includes(argTeamId)) {
            userConfig.teamIds.push(argTeamId);
        }
    }
    const questions = [];
    const teamsQuestion = {
        type: "input",
        name: "teamIds",
        message: "Team IDs (space delimited)",
        validate: value => {
            if (!/\S/.test(value) && userConfig.teamIds.length < 1) {
                return `The list of team IDs you entered is empty`;
            }
            return true;
        },
    };
    if (userConfig.teamIds.length > 0) {
        teamsQuestion.default = userConfig.teamIds.join(" ");
    }
    questions.push(teamsQuestion);
    const configPath = configuration_1.userConfigPath();
    if (!userConfig.token) {
        console.log(`
As part of the Atomist configuration, we need to create a GitHub
personal access token for you that will be used to authenticate with
the Atomist API.  The personal access token will have "read:org" and
"repo" scopes, be labeled as being for the "Atomist API", and will be
written to a file on your local machine.  Atomist does not retain the
token nor your GitHub username and password.
`);
        if (!argGitHubUser) {
            questions.push({
                type: "input",
                name: "user",
                message: "GitHub Username",
                validate: value => {
                    if (!/^[-.A-Za-z0-9]+$/.test(value)) {
                        return `The GitHub username you entered contains invalid characters: ${value}`;
                    }
                    return true;
                },
            });
        }
        if (!argGitHubPassword) {
            questions.push({
                type: "password",
                name: "password",
                message: "GitHub Password",
                validate: value => {
                    if (value.length < 1) {
                        return `The GitHub password you entered is empty`;
                    }
                    return true;
                },
            });
        }
        if (!argGitHubMfaToken) {
            questions.push({
                type: "input",
                name: "mfa",
                message: "GitHub 2FA Code",
                when: answers => {
                    const user = (argGitHubUser) ? argGitHubUser : answers.user;
                    const password = (argGitHubPassword) ? argGitHubPassword : answers.password;
                    return createGitHubToken(user, password)
                        .then(token => {
                        userConfig.token = token;
                        return false;
                    })
                        .catch(err => {
                        if (err.code === 401 && err.message) {
                            const msg = JSON.parse(err.message);
                            const mfaErr = "Must specify two-factor authentication OTP code.";
                            if (msg.message.indexOf(mfaErr) > -1) {
                                return true;
                            }
                        }
                        throw err;
                    });
                },
                validate: (value, answers) => {
                    if (!/^\d{6}$/.test(value)) {
                        return `The GitHub 2FA you entered is invalid, it should be six digits: ${value}`;
                    }
                    return true;
                },
            });
        }
    }
    else {
        console.log(`
Your Atomist client configuration already has an access token.
To generate a new token, remove the existing token from
'${configPath}'
and run \`atomist config\` again.
`);
    }
    if (process.env.ATOMIST_DOCKER_REGISTRY || process.env.ATOMIST_DOCKER_USER ||
        process.env.ATOMIST_DOCKER_PASSWORD) {
        const dockerRegistry = {
            type: "input",
            name: "dockerRegistry",
            message: "Docker Registry Host",
            validate: value => {
                if (!/\S/.test(value) && !process.env.ATOMIST_DOCKER_REGISTRY) {
                    return `The Docker registry host you entered is empty`;
                }
                return true;
            },
        };
        if (process.env.ATOMIST_DOCKER_REGISTRY) {
            dockerRegistry.default = process.env.ATOMIST_DOCKER_REGISTRY;
        }
        const dockerUser = {
            type: "input",
            name: "dockerUser",
            message: "Docker Registry Username",
            validate: value => {
                if (!/\S/.test(value) && !process.env.ATOMIST_DOCKER_USER) {
                    return `The Docker registry username you entered is empty`;
                }
                return true;
            },
        };
        if (process.env.ATOMIST_DOCKER_USER) {
            dockerUser.default = process.env.ATOMIST_DOCKER_USER;
        }
        const dockerPassword = {
            type: "password",
            name: "dockerPassword",
            message: "Docker Registry Password",
            validate: value => {
                if (!/\S/.test(value) && !process.env.ATOMIST_DOCKER_PASSWORD) {
                    return `The Docker registry password you entered is empty`;
                }
                return true;
            },
        };
        if (process.env.ATOMIST_DOCKER_PASSWORD) {
            dockerPassword.default = process.env.ATOMIST_DOCKER_PASSWORD;
        }
        questions.push(dockerRegistry, dockerUser, dockerPassword);
    }
    if (process.env.PCF_API || process.env.PCF_ORG || process.env.PCF_SPACE_STAGING ||
        process.env.PCF_SPACE_PRODUCTION || process.env.PIVOTAL_USER || process.env.PIVOTAL_PASSWORD) {
        const cfUser = {
            type: "input",
            name: "cfUser",
            message: "Cloud Foundry Username",
            validate: value => {
                if (!/\S/.test(value) && !process.env.PIVOTAL_USER) {
                    return `The Cloud Foundry username you entered is empty`;
                }
                return true;
            },
        };
        if (process.env.PIVOTAL_USER) {
            cfUser.default = process.env.PIVOTAL_USER;
        }
        const cfPassword = {
            type: "password",
            name: "cfPassword",
            message: "Cloud Foundry Password",
            validate: value => {
                if (!/\S/.test(value) && !process.env.PIVOTAL_PASSWORD) {
                    return `The Cloud Foundry password you entered is empty`;
                }
                return true;
            },
        };
        if (process.env.PIVOTAL_PASSWORD) {
            cfPassword.default = process.env.PIVOTAL_PASSWORD;
        }
        const cfOrg = {
            type: "input",
            name: "cfOrg",
            message: "Cloud Foundry Org",
            validate: value => {
                if (!/\S/.test(value) && !process.env.PCF_ORG) {
                    return `The Cloud Foundry Org you entered is empty`;
                }
                return true;
            },
        };
        if (process.env.PCF_ORG) {
            cfOrg.default = process.env.PCF_ORG;
        }
        const cfSpaceStaging = {
            type: "input",
            name: "cfSpaceStaging",
            message: "Cloud Foundry Staging Space",
            validate: value => {
                if (!/\S/.test(value) && !process.env.PCF_SPACE_STAGING) {
                    return `The Cloud Foundry Space you entered is empty`;
                }
                return true;
            },
        };
        if (process.env.PCF_SPACE_STAGING) {
            cfOrg.default = process.env.PCF_SPACE_STAGING;
        }
        const cfSpaceProd = {
            type: "input",
            name: "cfSpaceProd",
            message: "Cloud Foundry Production Space",
            validate: value => {
                if (!/\S/.test(value) && !process.env.PCF_SPACE_PRODUCTION) {
                    return `The Cloud Foundry Space you entered is empty`;
                }
                return true;
            },
        };
        if (process.env.PCF_SPACE_PRODUCTION) {
            cfOrg.default = process.env.PCF_SPACE_PRODUCTION;
        }
        const cfApi = {
            type: "input",
            name: "cfApi",
            message: "Cloud Foundry API endpoint",
            validate: value => {
                if (!/\S/.test(value) && !process.env.PCF_API) {
                    return `The Cloud Foundry API endpoint you entered is empty`;
                }
                return true;
            },
        };
        if (process.env.PCF_API) {
            cfApi.default = process.env.PCF_API;
        }
        else {
            cfApi.default = "https://api.run.pivotal.io";
        }
        questions.push(cfUser, cfPassword, cfOrg, cfSpaceStaging, cfSpaceProd, cfApi);
    }
    if (process.env.ATOMIST_NPM) {
        const npm = {
            type: "input",
            name: "npm",
            message: "NPM configuration (JSON stringified)",
            validate: value => {
                if (!/\S/.test(value) && !process.env.ATOMIST_NPM) {
                    return `The NPM configuration you entered is empty`;
                }
                return true;
            },
            default: process.env.ATOMIST_NPM,
        };
        questions.push(npm);
    }
    if (process.env.USE_CHECKSTYLE || process.env.CHECKSTYLE_PATH) {
        const checkstyle = {
            type: "confirm",
            name: "checkstyle",
            message: "Use checkstyle?",
        };
        if (process.env.USE_CHECKSTYLE) {
            checkstyle.default = process.env.USE_CHECKSTYLE;
        }
        const checkstylePath = {
            type: "input",
            name: "checkstylePath",
            message: "Absolute path to checkstyle jar",
            validate: value => {
                if (!value && !process.env.CHECKSTYLE_PATH) {
                    return `You must provide a path to the checkstyle jar`;
                }
                if (value && !/\.jar$/i.test(value)) {
                    return `The path to the checkstyle jar must end with '.jar'`;
                }
                return true;
            },
        };
        if (process.env.CHECKSTYLE_PATH) {
            checkstylePath.default = process.env.CHECKSTYLE_PATH;
        }
        questions.push(checkstyle, checkstylePath);
    }
    return inquirer.prompt(questions)
        .then(answers => {
        if (answers.teamIds) {
            userConfig.teamIds = answers.teamIds.split(/\s+/);
        }
        const sdm = {};
        if (answers.dockerRegistry) {
            sdm.docker = {
                registry: answers.dockerRegistry,
                user: answers.dockerUser,
                password: answers.dockerPassword,
            };
        }
        if (answers.cfUser) {
            sdm.cloudfoundry = {
                user: answers.cfUser,
                password: answers.cfPassword,
                org: answers.cfOrg,
                space: {
                    staging: answers.cfSpaceStaging,
                    production: answers.cfSpaceProd,
                },
                api: answers.cfApi,
            };
        }
        if (answers.npm) {
            try {
                sdm.npm = JSON.parse(answers.npm);
            }
            catch (e) {
                e.message = `Failed to parse NPM configuration as JSON: ${e.message}`;
                return Promise.reject(e);
            }
        }
        if (answers.checkstylePath) {
            sdm.checkstyle = {
                enable: answers.checkstyle,
                path: answers.checkstylePath,
            };
        }
        if (userConfig.sdm) {
            _.merge(userConfig.sdm, sdm);
        }
        else {
            userConfig.sdm = sdm;
        }
        if (!userConfig.token) {
            const user = (argGitHubUser) ? argGitHubUser : answers.user;
            const password = (argGitHubPassword) ? argGitHubPassword : answers.password;
            const mfa = (argGitHubMfaToken) ? argGitHubMfaToken : answers.mfa;
            return createGitHubToken(user, password, mfa)
                .then(token => {
                userConfig.token = token;
            });
        }
    })
        .then(() => configuration_1.writeUserConfig(userConfig))
        .then(() => {
        console.info(`Successfully created Atomist client configuration: ${configPath}`);
        return 0;
    }, err => {
        console.error(`Failed to create client configuration '${configPath}': ${stringify(err)}`);
        return 1;
    });
}
exports.cliAtomistConfig = cliAtomistConfig;
