// @atomist/microgrammar/spi/StringInputStream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Simple implementation of InputStream, backed by a string held in memory.
 */
var StringInputStream = /** @class */ (function () {
    /**
     * Create a new string-backed stream
     * @param content string
     * @param initialOffset initial offset. If supplied this pretends to the left
     * offset of the string. This enables us to match within matches
     */
    function StringInputStream(content, initialOffset) {
        if (initialOffset === void 0) { initialOffset = 0; }
        this.content = content;
        this.initialOffset = initialOffset;
        this.offset = 0;
        if (content === undefined) {
            throw new Error("Undefined content");
        }
    }
    StringInputStream.prototype.exhausted = function () {
        return this.offset - this.initialOffset >= this.content.length;
    };
    StringInputStream.prototype.read = function (n) {
        var s = this.content.substr(this.offset - this.initialOffset, n);
        this.offset += s.length;
        // console.log(`read(${n}) returned [${s}], offset now=${this.offset}`);
        return s;
    };
    return StringInputStream;
}());
exports.StringInputStream = StringInputStream;
