// @atomist/automation-client/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var decorators_1 = require("./decorators");
exports.CommandHandler = decorators_1.CommandHandler;
exports.ConfigurableCommandHandler = decorators_1.ConfigurableCommandHandler;
exports.EventHandler = decorators_1.EventHandler;
exports.MappedParameter = decorators_1.MappedParameter;
exports.MappedParameters = decorators_1.MappedParameters;
exports.Parameter = decorators_1.Parameter;
exports.Parameters = decorators_1.Parameters;
exports.Secret = decorators_1.Secret;
exports.Secrets = decorators_1.Secrets;
exports.Value = decorators_1.Value;
exports.Tags = decorators_1.Tags;
var HandlerResult_1 = require("./HandlerResult");
exports.failure = HandlerResult_1.failure;
exports.Failure = HandlerResult_1.Failure;
exports.FailurePromise = HandlerResult_1.FailurePromise;
exports.reduceResults = HandlerResult_1.reduceResults;
exports.success = HandlerResult_1.success;
exports.Success = HandlerResult_1.Success;
exports.SuccessPromise = HandlerResult_1.SuccessPromise;
const GraphQL = require("./graph/graphQL");
exports.GraphQL = GraphQL;
var logger_1 = require("./internal/util/logger");
exports.logger = logger_1.logger;
var ingesters_1 = require("./ingesters");
exports.buildEnum = ingesters_1.buildEnum;
exports.buildIngester = ingesters_1.buildIngester;
exports.buildType = ingesters_1.buildType;
exports.IngesterBuilder = ingesters_1.IngesterBuilder;
var atomistWebhook_1 = require("./atomistWebhook");
exports.postAtomistWebhook = atomistWebhook_1.postAtomistWebhook;
var globals_1 = require("./globals");
exports.automationClientInstance = globals_1.automationClientInstance;
