// @atomist/automation-client/server/AutomationEventListener.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class AutomationEventListenerSupport {
    registrationSuccessful(handler) {
        // This is intentionally left empty
    }
    contextCreated(context) {
        // This is intentionally left empty
    }
    commandIncoming(payload) {
        // This is intentionally left empty
    }
    commandStarting(payload, ctx) {
        // This is intentionally left empty
    }
    commandSuccessful(payload, ctx, result) {
        return Promise.resolve();
    }
    commandFailed(payload, ctx, err) {
        return Promise.resolve();
    }
    eventIncoming(payload) {
        // This is intentionally left empty
    }
    eventStarting(payload, ctx) {
        // This is intentionally left empty
    }
    eventSuccessful(payload, ctx, result) {
        return Promise.resolve();
    }
    eventFailed(payload, ctx, err) {
        return Promise.resolve();
    }
    messageSent(message, destinations, options, ctx) {
        // This is intentionally left empty
    }
}
exports.AutomationEventListenerSupport = AutomationEventListenerSupport;
