// @atomist/automation-client/operations/edit/editModes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function toEditModeFactory(em) {
    return p => isEditMode(em) ? em : em(p);
}
exports.toEditModeFactory = toEditModeFactory;
function isEditMode(em) {
    return !!em.message;
}
exports.isEditMode = isEditMode;
/**
 * Return a commit to master branch with the given message. Use with care!
 */
function commitToMaster(message) {
    return {
        message,
        branch: "master",
    };
}
exports.commitToMaster = commitToMaster;
function isBranchCommit(em) {
    const maybeBc = em;
    return !!maybeBc.branch && !!maybeBc.message;
}
exports.isBranchCommit = isBranchCommit;
/**
 * Captures extra steps that must go into raising a PR
 */
class PullRequest {
    constructor(branch, title, body = title, message = title) {
        this.branch = branch;
        this.title = title;
        this.body = body;
        this.message = message;
    }
}
exports.PullRequest = PullRequest;
function isPullRequest(em) {
    const maybePr = em;
    return !!maybePr.branch && !!maybePr.body && !!maybePr.title;
}
exports.isPullRequest = isPullRequest;
function isCustomExecutionEditMode(ei) {
    const maybeCei = ei;
    return !!maybeCei.edit;
}
exports.isCustomExecutionEditMode = isCustomExecutionEditMode;
