// @atomist/automation-client/operations/generate/NewRepoCreationParameters.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const decorators_1 = require("../../decorators");
const gitHubPatterns_1 = require("../common/params/gitHubPatterns");
/**
 * Parameters common to all generators that create new repositories
 */
class NewRepoCreationParameters {
    constructor() {
        this.description = "my new project";
        this.visibility = "public";
    }
}
__decorate([
    decorators_1.MappedParameter(decorators_1.MappedParameters.GitHubOwner),
    __metadata("design:type", String)
], NewRepoCreationParameters.prototype, "owner", void 0);
__decorate([
    decorators_1.Parameter({
        pattern: gitHubPatterns_1.GitHubNameRegExp.pattern,
        displayName: "Target Repository Name",
        description: "name of the target repository",
        validInput: gitHubPatterns_1.GitHubNameRegExp.validInput,
        minLength: 1,
        maxLength: 100,
        required: true,
        order: 1,
    }),
    __metadata("design:type", String)
], NewRepoCreationParameters.prototype, "repo", void 0);
__decorate([
    decorators_1.Parameter({
        displayName: "Project Description",
        description: "short descriptive text describing the new project",
        validInput: "free text",
        minLength: 1,
        maxLength: 100,
        required: false,
    }),
    __metadata("design:type", String)
], NewRepoCreationParameters.prototype, "description", void 0);
__decorate([
    decorators_1.Parameter({
        displayName: "Repository Visibility",
        description: "visibility of the new repository (public or private; defaults to public)",
        pattern: /^(public|private)$/,
        validInput: "public or private",
        minLength: 6,
        maxLength: 7,
        required: false,
    }),
    __metadata("design:type", String)
], NewRepoCreationParameters.prototype, "visibility", void 0);
exports.NewRepoCreationParameters = NewRepoCreationParameters;
